From 238cb1eb95deab1080f20b4def7db74a67890676 Mon Sep 17 00:00:00 2001
From: q66 <q66@chimera-linux.org>
Date: Sun, 28 Jul 2024 21:34:35 +0200
Subject: [PATCH] cgroups: set up controllers ahead of time when mounting

This is needed for container runtimes as enabling controllers for
the root cgroup may not be possible on-demand (particularly, the
cpu controller will not enable if there are any realtime-priority
processes around, and instead it will return EINVAL and that may
make container runtimes fail).

By enabling them all ahead of time, we guarantee that this will
succeed as only pid1 runs at that time.
---
 early/scripts/cgroups.sh | 20 ++++++++++++++++++--
 1 file changed, 18 insertions(+), 2 deletions(-)

diff --git a/early/scripts/cgroups.sh b/early/scripts/cgroups.sh
index 8389822..432a1d0 100644
--- a/early/scripts/cgroups.sh
+++ b/early/scripts/cgroups.sh
@@ -7,5 +7,21 @@ set -e
 
 . ./early/scripts/common.sh
 
-mkdir -p "/sys/fs/cgroup"
-./early/helpers/mntpt "/sys/fs/cgroup" || mount -t cgroup2 -o nsdelegate cgroup2 "/sys/fs/cgroup"
+CG_PATH="/sys/fs/cgroup"
+
+mkdir -p "$CG_PATH"
+./early/helpers/mntpt "$CG_PATH" || mount -t cgroup2 -o nsdelegate cgroup2 "/sys/fs/cgroup"
+
+# just in case
+[ -e "${CG_PATH}/cgroup.subtree_control" ] || exit 0
+[ -e "${CG_PATH}/cgroup.controllers" ] || exit 0
+
+# get the available controllers
+read -r CG_ACTIVE < "${CG_PATH}/cgroup.controllers"
+
+# enable them individually; if some fail, that's ok
+# we want to enable things here as it may not be possible later
+# (e.g. cpu will not enable when there are any rt processes running)
+for cont in ${CG_ACTIVE}; do
+    echo "+${cont}" > "${CG_PATH}/cgroup.subtree_control" 2>/dev/null || :
+done
